### PR TITLE
Refactor Graql EquivalentFragmentSets

### PR DESCRIPTION
**Pull request by aelred**
Add subclasses for every kind of equivalent fragment set in Graql. This is necessary so that we can recognise certain patterns of fragment sets in the queries and optimise them.

### CODE REVIEW

**Comment by aelred** [comment]
please retest this please - due to `SingleQueueTaskManagerTest`

File: grakn-core/src/main/java/ai/grakn/util/Schema.java [hunk 1]
```diff
@@ -263,4 +263,14 @@ public TypeName getName() {
...
         }
     }
+
+    /**
+     *
+     * @param typeName The resource type name
+     * @param value The value of the resource
+     * @return A unique id for the resource
+     */
+    public static String generateResourceIndex(TypeName typeName, String value){
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/util/Schema.java]: I am starting to think we dumping too much in `Schema`. Maybe we should create a `CoreUtils` ? 

In my view `Schema` should only contain our hardcoded enums and configs.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/CastingFragmentSet.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package ai.grakn.graql.internal.gremlin.sets;
+
+import ai.grakn.graql.VarName;
+import ai.grakn.graql.internal.gremlin.EquivalentFragmentSet;
+
+import static ai.grakn.graql.internal.gremlin.fragment.Fragments.inCasting;
+import static ai.grakn.graql.internal.gremlin.fragment.Fragments.outCasting;
+
+/**
+ * @author Felix Chapman
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/CastingFragmentSet.java]: -1 for bad javadocs

Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/CastingFragmentSet.java]: Yeah, I was torn on this, too... my reasoning was that these are package-private. The `public` factory methods to create these are in `EquivalentFragmentSets` - and that _is_ JavaDoc'd.

File: grakn-core/src/main/java/ai/grakn/util/Schema.java (hunk 1 above)
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/util/Schema.java]: I agree. I don't much like `CoreUtils` though - I'll think about it!

Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/util/Schema.java]: Bad name I admit but some place where we can dump these methods used throughout the modules would be handy.

**Pull request merged**